### PR TITLE
consistent ignore keys + make private

### LINKED ISSUE #7258: [save/load model] authorized keys, no save keys, etc.
We already have:

```
   # modeling_bart:
    authorized_missing_keys = [r"final_logits_bias", r"encoder\.version", r"decoder\.version"]
```
Once https://github.com/huggingface/transformers/pull/7224 is merged we will have:

```
class SomeModel(PretrainedFSMTModel):
    base_model_prefix = "model"
    authorized_missing_keys = [
        "model.encoder.embed_positions.weight",
        "model.decoder.embed_positions.weight",
    ]
    keys_to_never_save= [
        "model.encoder.embed_positions.weight",
        "model.decoder.embed_positions.weight",
    ]
```

I'd like to discuss several things:

1. let's pick consistent intuitive names for the group of these class variables - e.g. both ending with `_keys` or starting with `keys_` and use more descriptive mnemonics?

I suggest:

* `keys_to_ignore_on_load`
* `keys_to_ignore_on_save`

---------

2. why is the current implementation for `authorized_missing_keys` uses a regex search?

```
   # modeling_bart:
    authorized_missing_keys = [r"final_logits_bias", r"encoder\.version", r"decoder\.version"]
    [...]
   # modeling_utils:
            if cls.authorized_missing_keys is not None:
                for pat in cls.authorized_missing_keys:
                    missing_keys = [k for k in missing_keys if re.search(pat, k) is None]
```
when a simple direct comparison `k in list` would work just fine? i.e.: (untested)
```
if cls.authorized_missing_keys
    missing_keys = [k for k in missing_keys if k is not in cls.authorized_missing_keys]
```

it'd make the listing of the keys easier to write/read.
```
    authorized_missing_keys = ["final_logits_bias", "encoder.version", "decoder.version"]
```

-------------------

3. I think we may have an issue with `model.` prefix present in some saved model `state_dict`s and lacking in others - or is it a save/load issue where it has the prefix on the way out, but not back in? Note, that `authorized_missing_keys` doesn't have the `model.` prefix. Perhaps that's why the regex was used - to catch w/ and w/o the prefix?

So perhaps the normalization could happen in the core libs (load?) and the models listing special "needs" keys should all either have the prefix or not. 

----

Thank you for your input.

### PR DESCRIPTION
**Pull request by stas00**
This PR addresses https://github.com/huggingface/transformers/issues/7258

(the proposal has evolved a bit since the initial PR, this comments reflects the current state)

* [x] renames optional model attributes:
```
  - authorized_missing_keys    => _keys_to_ignore_on_load_missing
  - authorized_unexpected_keys => _keys_to_ignore_on_load_unexpected
  - keys_to_never_save         => _keys_to_ignore_on_save
```
to (1) make them consistent (2) make them private

* [x] removes these from public API docstring (documents them privately as comments in place)

This is a breaking change.

Fixes https://github.com/huggingface/transformers/issues/7258

@LysandreJik, @sgugger 

p.s. if we want to postpone it for v5, this PR was a quick one liner:
```
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|authorized_missing_keys|_keys_to_ignore_on_load_missing|g; s|authorized_unexpected_keys|_keys_to_ignore_on_load_unexpected|g'; s|keys_to_never_save|_keys_to_ignore_on_save|g' {} \;
```
and then manually adjusting the docs.

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -343,14 +343,14 @@ class TFPreTrainedModel(tf.keras.Model, TFModelUtilsMixin, TFGenerationMixin):
           :class:`~transformers.PretrainedConfig` to use as configuration class for this model architecture.
         - **base_model_prefix** (:obj:`str`) -- A string indicating the attribute associated to the base model in
           derived classes of the same architecture adding modules on top of the base model.
-        - **authorized_missing_keys** (:obj:`List[str]`, `optional`) -- A list of re pattern of tensor names to ignore
+        - **_keys_to_ignore_on_load** (:obj:`List[str]`, `optional`) -- A list of re pattern of tensor names to ignore
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:346]: I think we should stop documenting this since it's private.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -404,17 +404,17 @@ class PreTrainedModel(nn.Module, ModuleUtilsMixin, GenerationMixin):
 
         - **base_model_prefix** (:obj:`str`) -- A string indicating the attribute associated to the base model in
           derived classes of the same architecture adding modules on top of the base model.
-        - **authorized_missing_keys** (:obj:`Optional[List[str]]`) -- A list of re pattern of tensor names to ignore
+        - **_keys_to_ignore_on_load** (:obj:`Optional[List[str]]`) -- A list of re pattern of tensor names to ignore
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:407]: Same comment as above.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -404,17 +404,17 @@ class PreTrainedModel(nn.Module, ModuleUtilsMixin, GenerationMixin):
...
           when loading the model (and avoid unnecessary warnings).
-        - **keys_to_never_save** (:obj:`Optional[List[str]]`) -- A list of of tensor names to ignore when saving the
-          model (useful for keys that aren't trained, but which are deterministic)
+        - **_keys_to_ignore_on_save** (:obj:`Optional[List[str]]`) -- A list of of tensor names to ignore when saving
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:409]: Same

**Review by sgugger — APPROVED** [review]
Way cleaner, thanks! Just some doc nits on the side.

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (stas00) [reply src/transformers/modeling_tf_utils.py:346]: makes sense - but they should be documented somewhere else instead, no?

Comment (stas00) [reply src/transformers/modeling_tf_utils.py:346]: can there be a docstring section that is private?

Comment (stas00) [reply src/transformers/modeling_tf_utils.py:346]: regardless, if we want to get it into v4, then probably remove those as you suggested and then think of where to document those now private keys later?

**Review by LysandreJik — APPROVED** [review]
This is great, thanks for being so fast @stas00!

**Comment by stas00** [comment]
oh, boy, there is also `authorized_unexpected_keys`

**Comment by LysandreJik** [comment]
Indeed, very nice catch! How should we rename that one?

**Comment by stas00** [comment]
Current `authorized_missing_keys` and `authorized_unexpected_keys` do the same thing overall, just 2 different categories.

perhaps?

```
  - authorized_missing_keys    => _keys_to_ignore_on_load_missing
  - authorized_unexpected_keys => _keys_to_ignore_on_load_unexpected
  - keys_to_never_save         => _keys_to_ignore_on_save
```

**Comment by sgugger** [comment]
For me `authorized_unexpected_keys` should be the `_keys_to_ignore_on_load`: they are in the state dict but we ignore them.
The `authorized_missing_keys` should have another name such as `_keys_missing_to_ignore_on_load` or just `_keys_missing_to_ignore`.

Re- documentation. We usually document private stuff in comments in the code, so I think we should remove the public documentation and change it in comments.

**Comment by sgugger** [comment]
We were writing at the same time @stas00 , your names are better than mine. Go ahead!

File: src/transformers/modeling_tf_utils.py [hunk 4]
```diff
@@ -343,15 +343,15 @@ class TFPreTrainedModel(tf.keras.Model, TFModelUtilsMixin, TFGenerationMixin):
...
     config_class = None
     base_model_prefix = ""
-    authorized_missing_keys = None
-    authorized_unexpected_keys = None
+    # a list of re pattern of tensor names to ignore from the model when loading the model weights
+    # (and avoid unnecessary warnings).
+    _keys_to_ignore_on_load_missing = None
+    # a list of re pattern of tensor names to ignore from the weights when loading the model weights
+    # (and avoid unnecessary warnings).
+    _keys_to_ignore_on_load_unexpected = None
```
Comment (stas00) [inline src/transformers/modeling_tf_utils.py:354]: Is this good, @sgugger? I did the same for the other utils file but with 3 entries.

**Review by LysandreJik — APPROVED** [review]
I really like the new comments.

**Comment by stas00** [comment]
Ah, not sure why it doesn't let me rerun the CI - the failures are unrelated - failed connection.

**Comment by LysandreJik** [comment]
You can safely ignore the failed connections. It's been happening since the change to git-based repos. We're looking into fixing it with @julien-c, it happens very often.

**Comment by stas00** [comment]
@LysandreJik, I trust you will document this breaking change - I just don't know where I'd do that...

**Comment by LysandreJik** [comment]
Yes, I'm currently documenting all breaking changes in the release notes.
